app/models/project_group.rb

```diff
@@ -37,6 +37,26 @@ class DestroyWithChildrenError < StandardError; end
   # otherwise children are deleted before we even get here.
   before_destroy :validate_no_children, prepend: true
 
+  # Gets the total number of steps beneath this group.
+  # Currently this will recursively traverse the tree and fire a whole bunch of queries,
+  # one for each ProjectGroup. Should be some way to eager load but not seeing it.
+  # Performance shouldn't be too bad though as there shouldn't be that many groups.
+  def descendant_step_count
+    children.to_a.sum do |c|
+      c.is_a?(ProjectStep) ? 1 : c.descendant_step_count
+    end
+  end
+
+  # Determine if the group's children are all steps or a mix of steps and groups.
+  def descendants_only_steps?
+    children.each do |c|
+      if c.is_a?(ProjectGroup)
+        return false
+      end
+    end
+    return true
```

Comment:
I usually try to avoid checking the class type. Maybe instead of `is_a?` add two methods. Something like:

```ruby
# timeline_entry.rb
def is_step?
  false
end

# project_step
def is_step?
  true
end
```

It might be overkill, but it would make future refactors easier.